### DISCUSSION TITLE
Keycloak LDAP password

### DISCUSSION THREAD

**Discussion by gehrigmobi** [keycloak/keycloak#23024] [Q&A] [answered]
Hi, I'm using Keycloak 20.0.1 with OpenLDAP 2.6.3. I set up an ldap user federation and I'm able to query users and log in with clients via OpenID Connect using the `uid` and `userPassword` in ldap. I'm synching users, map some ldap attributes and the edit mode in keycloak is set to writeable. I can perfectly set ldap attributes from Keycloak.

However, if I want to change the password using Users->User details->Credentials->Set password, I get the error
> Error saving password: Could not modify attribute for DN [uid=testuser,ou=users,dc=example,dc=org]".

The log shows:
> 2023-09-06 18:55:00,553 WARN  [org.keycloak.services.resources.admin.UserResource] (executor-thread-507) Could not update user password.: org.keycloak.models.ModelException: Could not modify attribute for DN [uid=testuser,ou=users,dc=example,dc=org]
> at org.keycloak.storage.ldap.idm.store.ldap.LDAPOperationManager.modifyAttributes(LDAPOperationManager.java:609)
> at org.keycloak.storage.ldap.idm.store.ldap.LDAPIdentityStore.updateADPassword(LDAPIdentityStore.java:399)
> at org.keycloak.storage.ldap.idm.store.ldap.LDAPIdentityStore.updatePassword(LDAPIdentityStore.java:367)
> ...
> at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
> at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
> at java.base/java.lang.Thread.run(Thread.java:829)
> Caused by: javax.naming.directory.InvalidAttributeIdentifierException: [LDAP: error code 17 - unicodePwd: attribute type undefined]; remaining name 'uid=testuser,ou=users,dc=example,dc=org'

How can I configure Keycloak to use `userPassword` instead of `unicodePwd`? I do not see any mapper for the password.
What am I missing? Thank you for your help.

**Comment by justin-tay** [comment, marked as answer]
Keycloak will use `unicodePwd` if you configured the `Vendor` as `Active Directory`. You should be using `Other` instead if you are using OpenLDAP.

![ldap](https://github.com/keycloak/keycloak/assets/49700559/af2ed3a2-cab6-4af7-af57-0431055f44c6)

**Reply by gehrigmobi** [reply]
Thank you, this solved it!

**Reply by bilxio** [reply]
Thanks! this helps me a lot!
